jep/0000/README.adoc

```diff
@@ -0,0 +1,232 @@
+modules cannot be updated except part of the relatively slow Jenkins core weekly/LTS cycle.
+This makes it more difficult to keep up with libraries such as Apache SSH.
+
+== Reasoning
+
+=== Inlining all modules into core
+
+All of these modules could be made part of `jenkins-core.jar`.
+This would however go against the principle of allowing Jenkins to be modular
+and letting administrators and packaging systems (such as `jenkinsfile-runner`) decide which pieces are important.
+Some of them also have dependency issues (see below about `instance-identity`).
+
+=== Status quo
+
+These eight packages could be left as modules.
+But to reverse the question, suppose functionality of this kind,
+such as adding an SSH server,
+were proposed as a new feature of Jenkins.
+The author would likely be urged to do so as a new plugin.
+Jenkins core developers would certainly not accept a proposal to introduce a novel mechanism for bundling extensions
+when plugin packaging is used for everything else and works fine.
+
+You might assume that the existing modules were written as modules because they _had_ to be:
+they might have needed to be loaded in the same `ClassLoader` as `jenkins-core.jar`,
+for example to use `@MetaInfServices`.
+In fact this was not the case; they all implement regular `@Extension` points.
+
+=== Switching `groupId`
+
+The Maven group id could be switched to, say, `io.jenkins.plugins` to more closely match the conventional style.
+Instead the existing group and artifacts ids are retained (for example `org.jenkins-ci.modules:sshd`)
+to simplify updates from plugin POMs and to make it clear that these _are_ updates.
+
+=== Inlining `instance-identity` into core
+
+Since instance identity is conceptually a basic part of Jenkins core,
+and `InstanceIdentityProvider` expects to have a single implementation `InstanceIdentityRSAProvider`,
+it would be nicer to simply inline this into core and deprecate the extension point.
+That would also assure that `IdentityRootAction` is always present;
+and would ensure that inbound TCP agents always work, without needing to extract them to a plugin depending on `instance-identity`.
+(WebSocket-based inbound agents from link:../222/README.adoc[JEP-222] do not require `instance-identity`.)
+
+Unfortunately doing so would require including BouncyCastle in core,
+something we would prefer not do;
+it is currently included as a library wrapper plugin.
+For now it seems much more straightforward to make `instance-identity` a (detached) plugin.
+If this decision needs to be reversed in the future,
+`InstanceIdentityRSAProvider` or its equivalent could just be added to core
+and `instance-identity` made an empty deprecated “tombstone”.
+
+=== Keeping `ssh-cli-auth` separate
+
+As of the removal of Remoting-based CLI,
+`ssh-cli-auth` merely provided utility classes used by `sshd`,
+with no independent functionality.
+Therefore there is no advantage in offering it as a separate plugin.
```

Comment:
Do you purpose to join ssh-cli-auth-module to ssh-module?

Reply:
Yes, as in **Specification**.